Update Rita Official Version

Zalgo2462 opened the pull request:
We are poised to release the next version of RITA for the SANS 504 release!
While this pull request is open there will be a feature freeze on dev to allow testing, review, and debugging.
Please read through the changes, leave reviews, and comment!

# Changelog
---
### Improved Functionality
  * Added log hook for adding logs to Mongo
  * Added log hook for writing logs to the filesystem
    * Created `log` folder under `~/.rita`
  * Added the ability to parse Bro collection types
  * Remove unnecessary debugging output
  * Added the ability to analyze DNS logs
  * Fix hostnames to pull from DNS logs rather than HTTP logs
    * Moved hostnames table in configuation file URL section to DNS section
  * Removed GNU NetCat from config file
  * Implemented a new parser with bulk insertion to MongoDB
    * Added Import buffer option in the Bro config section
  * Fixed linter warnings

### New Functionality
  * Added show-user-agents command
  * Added html-report command
  * Added show-long-connections command
  * Added show-long-urls command
  * Added show-most-visited-urls command
  * Removed whitelisting and associated config section
    * Ensured all output is `grep`-able
  * Implemented Exploded DNS Analysis (subdomain counting)
    * Added DNS section to configuration file
    * Added Exploded DNS Table under the DNS section
  * Enable install on Security Onion and Ubuntu 14.04-16.04

### Configuration Updates
  * Added DNS section to configuration file
  * Moved hostnames table in configuation file URL section to DNS section
  * Added Exploded DNS Table under the DNS section
  * Added Import buffer option in the Bro config section

### Installer Updates
  * Updated installer to detect if Go is already installed
  * Updated installer to detect if Bro is already installed
  * Updated installer to detect if Git is already installed

### Bug Fixes
  * Fixed a crash bug caused by long lines in Bro Logs
  * Fixed a bug in retrieving connection logs
  * Prevent socket timeouts on long Mongo operations

### Documentation
  * Added Contributing Guide
  * Cleaned up Readme
  * Cleaned up Manual Installation
  * Added RITA Gittiquette Guide

Zalgo2462 commented on database/resources.go:
```diff
@@ -43,6 +44,9 @@ func InitResources(cfgPath string) *Resources {
 	}
+	session.SetSocketTimeout(2 * time.Hour)
+	session.SetSyncTimeout(2 * time.Hour)
+	session.SetCursorTimeout(0)
```
Prevent network errors from crashing RITA

Zalgo2462 commented on datatypes/data/data.go:
```diff
@@ -10,13 +10,30 @@ type (
-		Src      string        `bson:"id_orig_h,omitempty"`
-		Spt      int           `bson:"id_orig_p,omitempty"`
+		Src      string        `bson:"id_origin_h,omitempty"`
+		Spt      int           `bson:"id_origin_p,omitempty"`
```
s/id_orig_h/id_origin_h this was a bug that was in the code base since at least September of 2016

Zalgo2462 commented on install.sh:
```diff
@@ -77,18 +77,25 @@ __install() {
 	echo "
+[+] Ensuring git is installed...
+"
+	apt install -y git
```
... ubuntu doesn't come with git >.>

Zalgo2462 commented on parser/fileparser.go:
```diff
@@ -0,0 +1,297 @@
+		scanner = bufio.NewScanner(fileHandle)
+	}
+
+	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)
```
Fixed a bug where huge lines in http logs wouldn't fit in the scanners buffer.

Zalgo2462 commented on parser/fsimporter.go:
```diff
@@ -0,0 +1,277 @@
+}
+
+//Run starts importing a given path into a datastore
+func (fs *FSImporter) Run(datastore *MongoDatastore) {
```
Where the parsing process begins

Zalgo2462 commented on parser/parsetypes/conn.go:
```diff
@@ -52,29 +51,20 @@ type (
 		RespIPBytes int64 `bson:"resp_ip_bytes" bro:"resp_ip_bytes" brotype:"count"`
 		// TunnelParents lists tunnel parents
-		TunnelParents string `bson:"tunnel_parents" bro:"tunnel_parents" brotype:"set[string]"`
+		TunnelParents []string `bson:"tunnel_parents" bro:"tunnel_parents" brotype:"set[string]"`
```
Fixed a bug where tunnel parents in a conn log would kill RITA

Zalgo2462 commented on parser/parsetypes/conn.go:
```diff
@@ -52,29 +51,20 @@ type (
+//TargetCollection returns the mongo collection this entry should be inserted
+//into
+func (in *Conn) TargetCollection(config *config.StructureCfg) string {
+	return config.ConnTable
```
Actually name the base collections based on the config file

Zalgo2462 commented on parser/parsetypes/conn.go:
```diff
@@ -52,29 +51,20 @@ type (
 
-// GetHostName is our method for collecting host name
-// Temporary function
-func (in *Conn) IsWhiteListed(whitelist []string) bool {
```
Removed whitelisting. Use grep.

Zalgo2462 commented on parser/parsetypes/parsetypes.go:
```diff
@@ -7,100 +36,112 @@ package parser
 const (
 	// BOOL reflects true or false, designated 'T' or 'F'
-	BOOL = "bool"
+	Bool = "bool"
```
gometalinter dislikes CAP CONSTANTS

Zalgo2462 commented on parser/parsetypes/parsetypes.go:
```diff
@@ -7,100 +36,112 @@ package parser
+	StringVector = "vector[string]"
+
+	// INTERVAL_VECTOR is a VECTOR which contains INTERVALs
+	IntervalVector = "vector[interval]"
```
Read collections out of bro!

Zalgo2462 commented:
BUG: config flag help text: use CONFIGFILE when as configuration when running this command

Zalgo2462 commented:
BUG: 
![image](https://cloud.githubusercontent.com/assets/761220/26686663/e105bf84-46aa-11e7-9ee9-626997a339cb.png)

s/TS in/INTVL/ as in the console output

TODO: Add urls and useragent to html output.

Zalgo2462 commented:
Pics!
![image](https://cloud.githubusercontent.com/assets/761220/26698957/3747f7d4-46d5-11e7-8380-0e8f60d33057.png)
![image](https://cloud.githubusercontent.com/assets/761220/26698979/4be8157a-46d5-11e7-8b01-6e340f7bb7a2.png)

ethack commented:
I'm getting a build error:
```
➜  make
go get
analysis/beacon/beacon.go:13:2: case-insensitive import collision: "github.com/Sirupsen/logrus" and "github.com/sirupsen/logrus"
make: *** [default] Error 1
```
[That line](https://github.com/ocmdev/rita/blob/24bf23602e26ba5b2e70398bd8a09b707c0596b3/analysis/beacon/beacon.go#L19) has `github.com/Sirupsen/logrus` but it looks like the user has changed his name to lowercase `sirupsen`.  Is the best way to fix this to change the import to lowercase or is there some underlying issue that we can fix that will prevent third party repo case changes from breaking the build?

Zalgo2462 commented:
So it appears this is a long running issue. Sirupsen renamed his account to sirupsen, but left a redirect in place for the upper case account. If all of your code uses one or the other everything will work. In the dev branch, we have added a library which used the new lowercase name. This seemed to build fine during progressive building/testing. But on a clean install go sees Sirupsen and sirupsen and gets confused. I've changed all of our references to the lowercase version and forked one of our dependencies which used the uppercase version. In addition, I've issued a PR for that dependency for them to use the lowercase version.

JonZeolla commented:
Spun this up, install worked properly.  Will plan to do more detailed testing next week if this isn't already merged.

JonZeolla commented:
Did some importing and analyzation with info from publicly available PCAPs.  The only error I ran into was `ERRO[0049] Failed:  (Name error: check collection name in yaml file and config)` during `rita analyze`, which may just be a config issue on my end.  Right now I'm spinning up bro against some larger, malicious PCAPs and then will analyze.

Zalgo2462 commented:
The host names table moved from the URLs section to the dns section. I would guess that would be the issue if you're using your old config. I highly recommend copying the new config from etc/rita.yaml from the dev branch to ~/.rita/config.yaml. You can also use rita testconfig and look for any fields that are zero or empty string

JonZeolla commented:
I did some testing with the [maccdc PCAPs](https://download.netresec.com/pcap/maccdc-2012/) from 2012 on AWS.  I have a few questions/comments:
 - Is there a changelog somewhere that I couldn't find?  I'm looking for a list of new/modified features to test without having to sort through all of the commits/file changes.
 - Is the UI that you provided screenshots of closed source?  I think I recall that being a purchased add-on, but wasn't sure and don't see a way to spin it up.
 - If you pull down dev and run `sudo ./install.sh` to set it up, `go get` pulls from the `master` ocmdev branch.  I had to `cd ~/go/src/github.com/ocmdev/rita; git checkout dev;make install` in order to test `dev` properly.  This was what was causing my `ERRO[0049] Failed: (Name error: check collection name in yaml file and config)` issue from earlier - may be worth throwing something in the `Readme.md` to help out future testers?

Zalgo2462 commented:
A changelog will be posted shortly.

The UI is produced using the rita html-report which is completely free!

The installer will only work with the master branch. You took the correct actions for compiling dev. 
However, you will need to update the config file you are using to match the one in etc/rita.yaml. I've put a contributing.md in the dev branch to help out.

The output for blacklist is erroneous. That number is the number of lines that COULDN'T be parsed. It is on the docket to be fixed in rita-blacklist.

JonZeolla commented:
RE: `contributing.md` Awesome, yeah I saw that, thanks.  

I updated the yaml appropriately last night and ran some analysis successfully, so that's looking pretty good.  I will do some more testing on and off (including the `html-report`) and check out the changelog once you throw it in.  This is all great stuff so far.
